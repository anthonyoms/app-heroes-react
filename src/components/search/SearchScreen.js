import React from "react";
import { heroes } from "../../data/heroes";
import { useForm } from "../../hooks/useForm";
import { HeroCard } from "../heroes/HeroCard";
import { useNavigate, useLocation } from "react-router-dom";

export const SearchScreen = () => {
    const location = useLocation();
    console.log(location)
    
  const navigate = useNavigate();
  const [values, handleInputChange, reset] = useForm({
    searchText: "",
  });

  const { searchText } = values;
  const heroesFiltered = heroes;

  const handleSearch = (e) => {
    e.preventDefault();
    navigate(`?q=${searchText}`);
    reset();
  };

  return (
    <div>
      <h1>SearchScreen</h1>
      <hr />

      <div className="row">
        <div className="col-5 animate__animated animate__fadeIn">
          <h4>Search Form</h4>
          <hr />
          <form onSubmit={handleSearch}>
            <input
              type="text"
              name="searchText"
              value={searchText}
              onChange={handleInputChange}
              autoComplete="off"
              placeholder="Find your hero"
              className="form-control"
            />
            <button
              type="sunmit"
              className="btn mt-1 btn-block btn-outline-primary"
            >
              Search
            </button>
          </form>
        </div>
        <div className="col-7 animate__animated animate__fadeInRight">
          <h4>Results</h4>
          <hr />

          {heroesFiltered.map((hero) => (
            <HeroCard key={hero.id} {...hero} />
          ))}
        </div>
      </div>
    </div>
  );
};
