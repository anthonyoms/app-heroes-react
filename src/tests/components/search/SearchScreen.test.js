import { mount } from "enzyme";
import { MemoryRouter, Route, Routes } from "react-router";
import { SearchScreen } from "../../../components/search/SearchScreen";

const mockedUsedNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockedUsedNavigate,
}));


describe("Pruebas en <SearchScreen />", () => {
  test("debe de mostrarse correctamente con los valores por defecto", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search"]}>
        <Routes>
          <Route path="/search" element={<SearchScreen />} />
        </Routes>
      </MemoryRouter>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(".text-success").text().trim()).toBe("Search a hero");
  });

  test("debe de mostrar a Batman y el input con el valor del queryString", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search?q=batman"]}>
        <Routes>
          <Route path="/search" element={<SearchScreen />} />
        </Routes>
      </MemoryRouter>
    );

    expect(wrapper.find("input").prop("value")).toBe("batman");
    expect(wrapper).toMatchSnapshot();
  });

  test("debe de mostrar un error si no se encuentra el Hero", () => {
    const hero = "batman!123";
    const wrapper = mount(
      <MemoryRouter initialEntries={[`/search?q=${hero}`]}>
        <Routes>
          <Route path="/search" element={<SearchScreen />} />
        </Routes>
      </MemoryRouter>
    );

    expect(wrapper.find(".text-danger").text().trim()).toBe(
      `There is no a hero with: ${hero}`
    );
  });

  test('debe llamar el navigate con el parametro correspondiente', () => {
    const hero = "batman!123";
    const wrapper = mount(
      <MemoryRouter initialEntries={[`/search?q=${hero}`]}>
        <Routes>
          <Route path="/search" element={<SearchScreen />} />
        </Routes>
      </MemoryRouter>
    );

    wrapper.find('input').simulate('change', {
        target: {
            name: 'searchText',
            value:'batman'
        }
    });

    wrapper.find('form').prop('onSubmit')({
        preventDefault(){}
    });

    expect(mockedUsedNavigate).toHaveBeenCalledWith(`?q=batman`);
  });
  
});
