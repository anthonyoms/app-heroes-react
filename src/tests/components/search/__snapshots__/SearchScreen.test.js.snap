// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchScreen /> debe de mostrar a Batman y el input con el valor del queryString 1`] = `
<div>
  <h1>
    SearchScreen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5 animate__animated animate__fadeIn"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batman"
        />
        <button
          className="btn mt-1 btn-block btn-outline-primary"
          type="sunmit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7 animate__animated animate__fadeInRight"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="text-success"
      />
      <div
        className="card ms-3 animate__animated animate__fadeIn"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Batman"
              className="card-img"
              src="./assets/heroes/dc-batman.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Batman
              </h5>
              <p
                className="card-text"
              >
                Bruce Wayne
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                  Detective Comics #27
                </small>
              </p>
              <a
                href="/hero/dc-batman"
                onClick={[Function]}
              >
                Mas..
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en <SearchScreen /> debe de mostrarse correctamente con los valores por defecto 1`] = `
<div>
  <h1>
    SearchScreen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5 animate__animated animate__fadeIn"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          className="btn mt-1 btn-block btn-outline-primary"
          type="sunmit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7 animate__animated animate__fadeInRight"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="text-success"
      >
        Search a hero
      </div>
    </div>
  </div>
</div>
`;
